### ISSUE TITLE
Stack Overflow in JS client app

### ISSUE THREAD

**Issue by broguinn** [PostHog/posthog-js#1600]
I'm loading the posthog JS client via this script in the head:

```
<script>
    !function(t,e){var o,n,p,r;e.__SV||(window.posthog=e,e._i=[],e.init=function(i,s,a){function g(t,e){var o=e.split(".");2==o.length&&(t=t[o[0]],e=o[1]),t[e]=function(){t.push([e].concat(Array.prototype.slice.call(arguments,0)))}}(p=t.createElement("script")).type="text/javascript",p.crossOrigin="anonymous",p.async=!0,p.src=s.api_host.replace(".i.posthog.com","-assets.i.posthog.com")+"/static/array.js",(r=t.getElementsByTagName("script")[0]).parentNode.insertBefore(p,r);var u=e;for(void 0!==a?u=e[a]=[]:a="posthog",u.people=u.people||[],u.toString=function(t){var e="posthog";return"posthog"!==a&&(e+="."+a),t||(e+=" (stub)"),e},u.people.toString=function(){return u.toString(1)+".people (stub)"},o="init capture register register_once register_for_session unregister unregister_for_session getFeatureFlag getFeatureFlagPayload isFeatureEnabled reloadFeatureFlags updateEarlyAccessFeatureEnrollment getEarlyAccessFeatures on onFeatureFlags onSessionId getSurveys getActiveMatchingSurveys renderSurvey canRenderSurvey getNextSurveyStep identify setPersonProperties group resetGroups setPersonPropertiesForFlags resetPersonPropertiesForFlags setGroupPropertiesForFlags resetGroupPropertiesForFlags reset get_distinct_id getGroups get_session_id get_session_replay_url alias set_config startSessionRecording stopSessionRecording sessionRecordingStarted captureException loadToolbar get_property getSessionProperty createPersonProfile opt_in_capturing opt_out_capturing has_opted_in_capturing has_opted_out_capturing clear_opt_in_out_capturing debug".split(" "),n=0;n<o.length;n++)g(u,o[n]);e._i.push([i,s,a])},e.__SV=1)}(document,window.posthog||[]);
    posthog.init(
      <pk redacted>,
      {
        api_host:'https://us.i.posthog.com', 
        person_profiles: 'always',
        disable_session_recording: true,
        loaded: function(ph) {
    		var posthogId = posthog.get_distinct_id();
    		// Use the posthogId (persisted to local storage and cookie) to identify Sentry user.
    		if (posthogId && Sentry) {
    			Sentry.setUser({ id: posthogId, ip_address: "{{auto}}"});
    		}
            
            // When in staging, don't collect PostHog data
            if (window.location.href.includes("staging")) {
              // ph.opt_out_capturing(); // opts a user out of event capture
              ph.set_config({ disable_session_recording: true });
            }
    	}
      }
    );

</script>
```

Later, I make calls to `posthog.capture` and `posthog.isFeatureEnabled`.

I'm seeing an occasional stack overflow in the client that I suspect might break eventing and A/B testing.

```
type-utils.ts:20 Uncaught RangeError: Maximum call stack size exceeded
    at D (type-utils.ts:20:28)
    at posthog-core.ts:714:28
    at Array.forEach (<anonymous>)
    at g (index.ts:21:17)
    at ao._execute_array (posthog-core.ts:709:9)
    at ao.push (posthog-core.ts:771:14)
    at t.<computed> [as capture] (?utm_source=redirect_to_join_form&utm_campaign=lootboxes:195:174)
    at ao.<anonymous> (posthog-core.ts:739:39)
    at Array.forEach (<anonymous>)
    at g (index.ts:21:17)
```

I've seen this issue in MacOS Chrome and Windows Chrome.

Based on the sourcemaps, this looks like it's related to function instrumentation. Do you have any idea how I might go about resolving this issue? It's currently blocking our ability to run A/B tests on the platform.

**Comment by rafaeelaudibert** [comment]
Hey, @broguinn! We've seen some other issues raising the same problem, but yours is the first with an actual stacktrace, thank you! I will be investigating this during the day.

**Comment by broguinn** [comment]
@rafaeelaudibert Thanks! I've figured out a temporary solution that might help you out.

First, I tried removing the `loaded` function from the posthog.init config arg. This did not solve the issue.

I then dug around in this codebase and saw that `posthog.init` calls are deferred to after DCL, so tried this:

```js
  document.addEventListener('DOMContentLoaded', () => {
    posthog.init(
      <pk redacted>,
      {
        api_host:'https://charge.basepowercompany.com', 
        person_profiles: 'always',
        disable_session_recording: true,
      }
    );
  });
``` 

And the problem seems solved after ~10 test loads. Hope this helps.

**Comment by rafaeelaudibert** [comment]
The problem with initializing posthog only after the DOM has finished loading is that you might lose events that happened **before** you init'd it.

The problem is likely happening exactly before you sometime have events being triggered before our script has finished loading. You say you're calling `posthog.capture` and `posthog.isFeatureEnabled`. Do you have anything else that might be called **before** `posthog` has finished being instantiated? Also, if you can and don't mind, would you be willing to share the payloads you're passing to those functions?

**Comment by rafaeelaudibert** [comment]
Also, do you mind sharing what's your `posthog-js` version, @broguinn?

**Comment by broguinn** [comment]
I feel pretty confident that this code executes before any other references to the `posthog` global variable - but it's great to know that I need to wait for DCL. If you get the stack overflow issue solved on chrome, please let me know and I'll remove the browser event callback on my side (plus I'd love to know the root cause).

Best of luck!

**Comment by broguinn** [comment]
@rafaeelaudibert you did warn me, but I am seeing an odd effect that I suspect is related to my work-around that waits to init the posthog JS client until DCL.

When I open a new browser (or incognito window) and make a bucketing decision in a context that should always override to a specific variant, the value is initially the control, even when I wrap the check in `posthog.onFeatureFlags`. When I refresh the page, the correct value is returned.

**Comment by broguinn** [comment]
@rafaeelaudibert I'm embarrassed to say that this is likely on our side, and that we've sent you on a bit of a wild rabbit chase. We called `posthog.init` twice in our code - and removing it appears to solve all of the stack overflow issues above.

Apologies again for the wasted time here.
